### PR TITLE
Support multiple horizon nodes.

### PR DESCRIPTION
**Pull request by avi-kik**
Also includes automated test support.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -9,6 +9,11 @@ it claims it has.
 ## Build
 
 This application was developed using Go 1.9, though earlier versions may work as well.
+Go expects a specific folder formation: /LOCAL_PATH/work/go/src/github.com/kinfoundation
```
Comment (oryband) [inline README.md]: kinecosystem instead of kinfoundation

File: cmd/loadtest/main.go [hunk 2]
```diff
@@ -25,9 +25,22 @@ import (
 // ClientTimeout is the Horizon HTTP request timeout.
 const ClientTimeout = 120 * time.Second
 
+// Support arrays of string flags
+type arrayFlags []string
+
+func (i *arrayFlags) String() string {
+	return "my string representation"
```
Comment (oryband) [inline cmd/loadtest/main.go]: you should print the array's contents here. otherwise if you print this object it will always output "my string representation"

File: cmd/loadtest/main.go [hunk 3]
```diff
@@ -25,9 +25,22 @@ import (
...
+}
+
+func (i *arrayFlags) Set(value string) error {
+	*i = append(*i, value)
+	return nil
+}
+
+var horizonDomainFlags arrayFlags
+
 var (
```
Comment (oryband) [inline cmd/loadtest/main.go:41]: add arrayflags to the list below

File: cmd/loadtest/main.go [hunk 4]
```diff
@@ -71,12 +86,15 @@ func Run() int {
...
 	}
 
-	client := horizon.Client{
-		URL:  *horizonDomainFlag,
-		HTTP: &http.Client{Timeout: ClientTimeout},
-	}
+	var clients []horizon.Client
+	for i := 0; i < len(horizonDomainFlags); i++ {
+		client := horizon.Client{
+			URL:  horizonDomainFlags[0],
```
Comment (oryband) [inline cmd/loadtest/main.go]: guess you wanted to use `[i]` instead of `[0]`?

File: cmd/loadtest/main.go [hunk 5]
```diff
@@ -43,6 +56,8 @@ var (
 
 // Run is the main function of this application. It returns a status exit code for main().
 func Run() int {
```
Comment (oryband) [inline cmd/loadtest/main.go:62]: i think that the best way to design this entire app is to do the following:
1. have a helper applicatino to pre-generate all transaction XDRs. similiar to the other app which creates acconuts beforehand.
1. load testing up will just load as much as necessary and submit them when the time comes

this will allow us to cache the premade txs instead of generating them as the test goes on (which takes time).

File: cmd/loadtest/main.go [hunk 6]
```diff
@@ -93,10 +111,10 @@ func Run() int {
 
 	// Generate workers for submitting operations.
 	submitters := make([]*submitter.Submitter, *numSubmittersFlag)
-	sequenceProvider := sequence.New(&client, logger)
+	sequenceProvider := sequence.New(&clients[0], logger)
```
Comment (oryband) [inline cmd/loadtest/main.go:116]: maybe generate the sequence provider it's own client?

File: cmd/loadtest/main.go [hunk 7]
```diff
@@ -71,12 +86,15 @@ func Run() int {
 		return 1
 	}
 
-	client := horizon.Client{
-		URL:  *horizonDomainFlag,
-		HTTP: &http.Client{Timeout: ClientTimeout},
-	}
+	var clients []horizon.Client
```
Comment (oryband) [inline cmd/loadtest/main.go:91]: needs to be a `[]*horizon.Client` since `horizon.Client` holds inner complex structs and not just primitives. apply to rest of code

File: cmd/loadtest/submitter/submitter.go [hunk 8]
```diff
@@ -140,7 +146,7 @@ func (s *Submitter) submit(logger log.Logger, destIndex int, native bool) error
 			amount = build.CreditAmount{"KIN", "GBSJ7KFU2NXACVHVN2VWQIXIV5FWH6A7OIDDTEUYTCJYGY3FJMYIDTU7", s.transferAmount}
 		}
 
-		ops = append(ops, build.Payment(build.Destination{AddressOrSeed: s.destinationAddresses[destIndex].Address()}, amount))
+	        ops = append(ops, build.Payment(build.Destination{AddressOrSeed: s.destinationAddresses[destIndex].Address()}, amount))
```
Comment (oryband) [inline cmd/loadtest/submitter/submitter.go]: this line doesn't change anything. what happened here?

File: cmd/loadtest/submitter/submitter.go [hunk 9]
```diff
@@ -120,7 +126,7 @@ func (s *Submitter) StartSubmission(ctx context.Context, limiter *rate.Limiter,
 // The transaction has the same property on every call:
 // Same source and and desitnation addresses, and same amount.
 // The only property that changes is the sequence number.
-func (s *Submitter) submit(logger log.Logger, destIndex int, native bool) error {
+func (s *Submitter) submit(logger log.Logger, destIndex int, native bool, clientIndex int) error {
```
Comment (oryband) [inline cmd/loadtest/submitter/submitter.go:129]: why not just transfer a reference to the client object i.e. `*horizon.Client` instead of its index?

File: cmd/python/prepare_accounts.sh [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+#AMOUNT_OF_SOURCE_ACCOUNTS=
+#AMOUNT_OF_DESTINATION_ACCOUNTS=
+#PASSPHRASE=$3
+#HORIZON=$4
+#FUNDER_SEED=$5
+echo $CHANNEL_SEED > channel_seed
+PASS=$PASSPHRASE
+
+pipenv sync; pipenv run python create.py --channel-seeds-file channel_seed --accounts $AMOUNT_OF_SOURCE_ACCOUNTS --passphrase "$PASS" --horizon $HORIZON --source-account $FUNDER_SEED --json-output True >../../accounts.json
+pipenv run python create.py --channel-seeds-file channel_seed --accounts $AMOUNT_OF_DESTINATION_ACCOUNTS --passphrase "$PASS" --horizon $HORIZON --source-account $FUNDER_SEED --json-output True >../../dest.json
```
Comment (oryband) [inline cmd/python/prepare_accounts.sh:18]: the `resources/create` and other binaries shouldn't be committed to this PR. they should be built by the `Makefile`. remove them

File: scripts/core_stats.sh [hunk 11]
```diff
@@ -0,0 +1,39 @@
+zgrep 'Advancing LCL: \[seq=[0-9]*' $1 | \
```
Comment (oryband) [inline scripts/core_stats.sh:1]: what is this script for? can you add some comments?

File: scripts/ingestion.sh [hunk 12]
```diff
@@ -0,0 +1,11 @@
+while read line; do
```
Comment (oryband) [inline scripts/ingestion.sh:1]: add short explanation explaining what this script does. also for other scripts

**Review by oryband — CHANGES_REQUESTED** [review]

**Comment by oryband** [comment]
also @avi-kik @ittiel there are a lot of commits here with the exact same commit message. squash them together or give them meaningful messages the diffrentiate them

File: cmd/python/prepare_accounts.sh (hunk 10 above)
Comment (ittiel) [reply cmd/python/prepare_accounts.sh:18]: As a concept I agree, these artifacts should be created as part of a CI flow, should be versioned, and saved in an artifactory.
But... it was not the scope of this task
I will add it as part of a different Jira epic of the CI

Comment (oryband) [reply cmd/python/prepare_accounts.sh:18]: i don't like adding binaries to a repo. cloning it afterwards takes a much longer time because the binaries need to be downloaded. i'd rather we'd build the required apps at runtime instead of committing them

File: cmd/loadtest/main.go (hunk 2 above)
Comment (avi-kik) [reply cmd/loadtest/main.go]: I copy+pasted that code as a hack for array support for flags.  I don't actually care about useful help messages here.

File: cmd/loadtest/main.go (hunk 4 above)
Comment (avi-kik) [reply cmd/loadtest/main.go]: Yes.

File: cmd/loadtest/main.go (hunk 5 above)
Comment (avi-kik) [reply cmd/loadtest/main.go:62]: I am considering writing a new app in Swift.  The only people who actually understand go-lang are too busy to maintain this one, and the changes we need are outgrowing my limited ability.

File: cmd/loadtest/submitter/submitter.go (hunk 8 above)
Comment (avi-kik) [reply cmd/loadtest/submitter/submitter.go]: I can only guess whitespace change, but I don't know.

File: cmd/loadtest/submitter/submitter.go (hunk 9 above)
Comment (avi-kik) [reply cmd/loadtest/submitter/submitter.go:129]: Because I don't know go, so I hack to make it work.

File: cmd/python/prepare_accounts.sh (hunk 10 above)
Comment (avi-kik) [reply cmd/python/prepare_accounts.sh:18]: In this case, building would take longer, as dependencies have to be downloaded.  Then we have the problem that we can't actually build it anyway, except on my (non-work) virtual linux machine.

File: scripts/core_stats.sh (hunk 11 above)
Comment (avi-kik) [reply scripts/core_stats.sh:1]: It's part of log collection.  There are several script fragments of this nature.

**Comment by avi-kik** [comment]
If we bother merging this, we would squash and merge.

**Comment by oryband** [comment]
taking over this PR. might break it to multiple branches and PRs for readability

**Comment by oryband** [comment]
I squashed all repeating commit messages and added a few small fixes following my review. merging this pr and continuing work in a new branch.
